example/src/main/java/io/netty/example/http2/tiles/http2/Http2RequestHandler.java

```diff
@@ -0,0 +1,138 @@
+ * which has been converted by a {@link InboundHttp2ToHttpAdapter} before it
+ * arrived here. For further details, check {@link Http2OrHttpHandler} where the
+ * pipeline is setup.
+ */
+public class Http2RequestHandler extends SimpleChannelInboundHandler<FullHttpRequest> {
+
+    static final String LATENCY = "latency";
+    static final String Y = "y";
+    static final String X = "x";
+    static final String PATH = "/http2";
+    static final AsciiString CONTENT_TYPE = new AsciiString("Content-Type");
+    static final byte[] dummy = "nothing".getBytes();
+
+    @Override
+    protected void channelRead0(ChannelHandlerContext ctx, FullHttpRequest request) throws Exception {
+
+        QueryStringDecoder queryString = new QueryStringDecoder(request.uri());
+        String streamId = streamId(request);
+
+        // Check arguments: path must match and latency parameter must be
+        // present
+        if (!PATH.equals(queryString.path()) || missing(queryString, LATENCY)) {
+            sendDummy(ctx, streamId);
+            return;
+        }
+
+        int latency = latency(queryString);
+
+        if (missing(queryString, X) && missing(queryString, Y)) {
+            handlePage(ctx, streamId, latency);
+        } else {
+            handleImage(queryString, ctx, streamId, latency);
+        }
+    }
+
+    private void handleImage(QueryStringDecoder query, ChannelHandlerContext ctx, String streamId, int latency) {
+        int x = toInt(query.parameters().get(X).get(0), 0);
+        int y = toInt(query.parameters().get(Y).get(0), 0);
+        byte[] image = ImageCache.instance().image(x, y);
+        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, OK, wrappedBuffer(image));
+        response.headers().set(CONTENT_TYPE, "image/jpeg");
+        sendResponse(ctx, streamId, latency, response);
+    }
+
+    private void handlePage(ChannelHandlerContext ctx, String streamId, int latency) {
+        ByteBuf content = Unpooled.wrappedBuffer(Html.HEADER, Html.body(latency), Html.FOOTER);
+        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, OK, content);
+        response.headers().set(CONTENT_TYPE, "text/html; charset=UTF-8");
+        sendResponse(ctx, streamId, latency, response);
+    }
+
+    protected void sendResponse(final ChannelHandlerContext ctx, String streamId, int latency,
+            final FullHttpResponse response) {
+        setContentLength(response, response.content().readableBytes());
+        setStreamId(response, streamId);
+        ctx.executor().schedule(new Runnable() {
+            @Override
+            public void run() {
+                ctx.writeAndFlush(response);
+            }
+        }, latency, TimeUnit.MILLISECONDS);
+    }
+
+    private int latency(QueryStringDecoder queryString) {
+        int latency = toInt(queryString.parameters().get(LATENCY).get(0), 0);
+        // Make sure that latency is not too big
+        return abs(latency > 1000 ? 0 : latency);
```

Comment:
Never, actually. It's just to prevent people from adding negative or too big values to the URL. That's potentially not needed, what happens if I schedule a task with a negative value?